marin/rl/envs/math_env.py

```diff
@@ -0,0 +1,224 @@
+            # Sample a problem (reuse the dataset when exhausted).
+            # ------------------------------------------------------------------
+            if self._example_idx >= len(self._examples):
+                self._rng.shuffle(self._examples)
+                self._example_idx = 0
+            example = self._examples[self._example_idx]
+            self._example_idx += 1
+
+            user_prompt: str = example["prompt"]
+            gt_answer: str = example["answer"]
+
+            # ------------------------------------------------------------------
+            # Call inference server.
+            # ------------------------------------------------------------------
+            completion = self._client.chat.completions.create(
+                model=self._model,
+                messages=[
+                    {"role": "user", "content": user_prompt},
+                ],
+            )
+
+            assistant_msg: str = completion.choices[0].message.content
+
+            # ------------------------------------------------------------------
+            # Reward calculation.
+            # ------------------------------------------------------------------
+            parsed = validate_format(assistant_msg + ">")  # util expects trailing >
+            is_valid = parsed["is_valid"]
+            extracted_answer = parsed["answer"]
+            is_correct = grade_answer(extracted_answer, gt_answer) if is_valid else False
+            reward = float(is_correct)
+
+            # ------------------------------------------------------------------
+            # Build rollout & emit.
+            # ------------------------------------------------------------------
+            turns = [
+                Turn(
+                    message=user_prompt,
+                    role="user",
+                    logprobs=None,
+                    reward=None,
+                    inference_metadata={},
+                ),
+                Turn(
+                    message=assistant_msg,
+                    role="assistant",
+                    logprobs=None,  # logprobs not available via OpenAI client
+                    reward=reward,
+                    inference_metadata={
+                        "model": self._model,
+                        "finish_reason": completion.choices[0].finish_reason,
+                    },
+                ),
+            ]
+            rollout = Rollout(turns=turns, metadata={"problem": user_prompt})
+            group = RolloutGroup(
+                id=f"math-{iteration}",
+                source="math_env",
+                created=time.time(),
+                rollouts=[rollout],
+                metadata={"valid_format": is_valid, "correct": is_correct},
+            )
+
+            # Push to learner / sink.
+            self._send_rollouts(group)
+
+            iteration += 1
+            await asyncio.sleep(0)  # yield control to Ray scheduler
```

Comment:
This seems like boilerplate and potentially easy to screw up - what if we made a default implementation of run in AbstractEnv which calls a `run_step`, which is just a normal function that returns `RolloutGroup`s?

Reply:
good idea. next PR?